packages/core/src/lib/filter/subscription.ts

```diff
@@ -0,0 +1,286 @@
+      throw new Error(
+        `Invalid event type. Expected content topic which should be a string but received ${contentTopic} of the type ${typeof contentTopic}`
+      );
+
+    const decoder = this.decoders.get(contentTopic);
+    if (!decoder)
+      throw new Error("No decoder found for content topic " + contentTopic);
+    if (decoder.pubsubTopic !== this.pubsubTopic)
+      throw new Error(
+        `Invalid decoder received. Expected decoder for pubsub topic ${this.pubsubTopic} but received decoder for pubsub topic ${decoder.pubsubTopic}`
+      );
+
+    super.addEventListener(contentTopic, listener, options);
+  }
+
+  async processMessage(message: WakuMessage): Promise<void> {
+    const { contentTopic } = message;
+    if (!contentTopic) {
+      log.info("Message has no content topic, skipping");
+      return;
+    }
+
+    const decoder = this.decoders.get(contentTopic);
+    if (decoder) {
+      try {
+        const decodedMessage = await decoder.fromProtoObj(
+          this.pubsubTopic,
+          message as IProtoMessage
+        );
+        if (!decodedMessage) {
+          throw new Error("Decoding failed");
+        }
+        this.dispatchEvent(
+          new CustomEvent<IDecodedMessage>(decoder.contentTopic, {
+            detail: decodedMessage
+          })
+        );
+      } catch (e) {
+        log.error("Error decoding message", e);
+      }
+    } else {
+      log.error("No decoder found for content topic", contentTopic);
+    }
+  }
+
+  private async sendSubscribeRequest(
+    request: FilterSubscribeRpc
+  ): Promise<FilterSubscribeResponse> {
+    const stream = await this.newStream(this.peer);
+    const res = await pipe(
+      [request.encode()],
+      lp.encode,
+      stream,
+      lp.decode,
+      async (source) => await all(source)
+    );
+
+    if (!res || !res.length) {
+      throw Error(
+        `No response received for request ${request.requestId}: ${res}`
+      );
+    }
+
+    return FilterSubscribeResponse.decode(res[0].slice());
+  }
+
+  private processSubscribeResponse<T extends IDecodedMessage>(
+    response: FilterSubscribeResponse,
+    decoders: IDecoder<T> | IDecoder<T>[]
+  ): void {
+    const { statusCode, requestId, statusDesc } = response;
+
+    if (statusCode < 200 || statusCode >= 300) {
+      throw new Error(
+        `Filter subscribe request ${requestId} failed with status code ${statusCode}: ${statusDesc}`
+      );
+    }
+
+    log.info(
+      "Subscribed to peer ",
+      this.peer.id.toString(),
+      "for content topics",
+      this.getContentTopics(decoders)
+    );
+  }
+
+  private getContentTopics<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[]
+  ): ContentTopic[] {
+    const decodersArray = Array.isArray(decoders) ? decoders : [decoders];
+    const decodersGroupedByCT = groupByContentTopic(decodersArray);
+    return Array.from(decodersGroupedByCT.keys());
+  }
+
+  private setNewDecoders<T extends IDecodedMessage>(
```

Comment:
Will there ever be a case where an existing decoder should be replaced?

Reply:
Interesting, I didn't think someone would need to replace decoders - but maybe adding decoders to the one already set.